### ISSUE TITLE
Future of this project

### ISSUE THREAD

**Issue by darosior** [wizardsardine/async-hwi#16]
It would be nice if we could make this project more friendly to downstream users. At the moment it is tailored to the needs of the [Liana](https://github.com/wizardsardine/liana) GUI, but i think a HWI library in a compiled language is dearly needed in the Bitcoin space. And Rust seems like a good fit for this. Plus, we already have the implementation for the Ledger and Specter devices.

I [tweeted about this](https://github.com/wizardsardine/liana) and it looks like there is interest from other Bitcoin developers in contributing to a Rust HWI. Given the interest in using or contributing to this library, i'm happy for Wizardsardine to dedicate some resources to extend the scope of this library to a wider audience than the Liana project. Let's discuss on this issue the path forward.

Regarding the goal of the library i think there isn't much to discuss. It should aim to be a high-level interface to the largest reasonable number of Bitcoin signing devices. I would just add a requirement that output descriptors be first class citizens. I'm not sure yet how (maybe using @bigspider's wallet policies?).

Regarding the path toward this goal, i think we:
1. First need to cleanup the current code. It's been a one man (@edouardparis)'s project and it could use some review (probably mine but anyone's welcome) / bit more testing.
2. Make the `async`ness optional.
3. Re-design the main interface, the `HWI` trait, with the new goals of this library in mind.
4. Add support for more signing devices.
5. Provide Python bindings.
6. Get to feature parity with the existing Python HWI.
7. ?Get the existing Python HWI to migrate to Rust HWI under the hood?
8. ?Honour an MSRV and move this repo to an org that isn't our company? Rust-Bitcoin maybe?

Thoughts? People interested in contributing to or using this Rust-HWI please chime in!

**Comment by vnprc** [comment]
It would be amazing to be able to compile to WASM to enable browser-based wallets.

**Comment by darosior** [comment]
Yes, it's something we are looking at as well. But there is more to it than simply compiling to WASM: the transport should be adapted as well, probably to something like web-HDI.

**Comment by edouardparis** [comment]
`navigator.hid.requestDevice()` and other browser javascript interfaces return js promise which are better handled by `wasm_bindgen_futures` async methods (ex: https://github.com/digitalbitbox/bitbox-api-rs/blob/master/src/wasm/connect.rs#L14)
WASM is one example of transport where the async feature has good support for, an other would be bluetooth as illustrated by [this crate](https://github.com/ewoolsey/ledger-bluetooth/tree/master/ledger-transport-ble) that implements the bluetooth communication for ledger Nano X. In fact,  Rust hardware ecosystem loves async, new exciting libraries like https://github.com/embassy-rs/embassy/ make full usage of it. This [tweet](https://twitter.com/NoRadsGiven/status/1592197489494724609) explains one of the reason.

I believe if we want to be ambitious in terms of feature and diversity of transport for this library, we should keep the async support and keep the name crate `async-hwi`. In the same time I totally understand the interest in a non-async library but I believe it would be too much code duplication and complexity (both in features flag and dependency management) to keep it in the same crate. If some individuals wants to start this project, they will not have to start from scratch as our library is a good way to understand what is possible in terms of common API between all the referenced devices. 

---
(Quick divergence on async rust, https://without.boats/blog/why-async-rust/ is a good read, in fact every article from https://without.boats/ are very good and should be read by every developers before screaming about coloring function)

**Comment by Sjors** [comment]
Another approach could be to add a command-line wrapper that behaves the same as the Python `hwi` command. That way it's a drop-in replacement that Bitcoin Core can call via the GUI and RPC (`enumeratesigners`, `signerdisplayaddress`, `send`, etc).

**Comment by pythcoiner** [comment]
there is already a CLI, but the API differ from `hwi`:

https://github.com/wizardsardine/async-hwi/tree/master/cli

**Comment by Sjors** [comment]
Thanks, that at least makes it easier to test manually.

**Comment by Sjors** [comment]
> make Core's HWI just be a simple Python frontend for that

Or make a compatible Rust frontend, I don't know how easy those are to package as a downloadable binary these days?

**Comment by trevarj** [comment]
Hi @Sjors, 

@edouardparis pointed out to me that you're working on some [features for this library](https://github.com/Sjors/async-hwi/pull/1) on your own fork.

Just letting you (and others here) know that I've been doing some work over on https://github.com/wizardsardine/bhwi/ if you are interested in having a look at it. It's not quite a drop-in replacement, but I think the architecture is very nice and extensible. It also has a lot of versatility for integration into other projects, since it is transport agnostic. It does not use any third-party device crates, like rust-coldcard or ledger's bitcoin_client_rs.

I've been trying to reach parity with the Python HWI interface/commands (as I said, not necessarily a drop-in). I've done a bunch of work to make the currently supported devices testable with emulators so those without the devices can help with implementation or testing. I've implemented BIP388 on rust-miniscript too, I'm using for displaying addresses and descriptor registration (in progress).

Please take a look and let us know if you have any questions. Would be great if bhwi can be improved to meet your goals, and then we won't have to overlap any work.

**Comment by Sjors** [comment]
@trevarj thanks, I wasn't aware of that.

I think what I'll do is first continue with https://github.com/Sjors/async-hwi/pull/2 for a bit, and then see if I can implement the same on `bhwi`. Then just compare the approach.

**Comment by Sjors** [comment]
@trevarj would you say the best way to use bhwi for an HWI clone, is to include it as a library in a new project?

**Comment by edouardparis** [comment]
I do not mind to add either `hwi` as a workspace member of  the bhwi repository or do
```
bhwi-cli/
├── Cargo.toml
└── src/
    └── bin/
        ├── bhwi.rs    # cli with commands following a modern convention like the one of `stevenroose/hal`
        └── hwi.rs     # cli reproducing the python hwi commands
```

as bhwi cli and hwi will have a lot of internal code in common.

**Comment by Sjors** [comment]
@edouardparis I'll let you or someone else take a stab at this. I plan to focus on the Bitcoin Core side of the required MuSig2 changes for the time being... not before I vibe code ColdCard support though :-)
